cli4/cli4.man
.TH CLI4 1

.SH NAME
cli4 \- Command line access to CloudFlare v4 API

.SH SYNOPSIS
.B cli4
[\fB\-V\fR]
[\fB\-h\fR|\fB\-\-help]
[\fB\-v\fR|\fB\-\-verbose]
[\fB\-q\fR|\fB\-\-quiet]
[\fB\-j\fR|\fB\-\-json]
[\fB\-y\fR|\fB\-\-yaml]
[\fBitem\fR=\fIvalue\fR ...]
[\fB\-G\fR|\fB\-\-get]
[\fB\-P\fR|\fB\-\-patch]
[\fB\-O\fR|\fB\-\-post]
[\fB\-U\fR|\fB\-\-put]
[\fB\-D\fR|\fB\-\-delete]
.IR /command ...

.SH DESCRIPTION
.B cli4
provides command line access to CloudFlare v4 API

.SH OPTIONS
.TP
.IP "[\-V, \-\-version]"
Display program version number and exit.
.IP "[\-h, \-\-help]"
This information (in a terse form).
.IP "[\-v, \-\-verbose]"
Provide some protcol debugging information.
.IP "[\-q, \-\-quiet]"
Don't output any JSON/YAML responses.
.IP "[\-j, \-\-json]"
Output response data in JSON format (the default).
.IP "[\-y, \-\-yaml]"
Output response data in YAML format (if yaml package installed).
.IP "\-\-get"
Send HTTP request as a \fBGET\fR (the default).
.IP "\-\-patch"
Send HTTP request as a \fBPATCH\fR.
.IP "\-\-post"
Send HTTP request as a \fBPOST\fR.
.IP "\-\-put"
Send HTTP request as a \fBPUT\fR.
.IP "\-\-delete"
Send HTTP request as a \fBDELETE\fR.
.IP "item=\fIvalue\fR"
Set a paramater or data value to send with a \fBGET\fR, \fBPATCH\fR, \fBPOST\fR, \fBPUT\fR or \fBDELETE\fR command. The value is sent as a string.
.IP item:=\fIvalue\fR
Set a paramater or data value to send with a \fBGET\fR, \fBPATCH\fR, \fBPOST\fR, \fBPUT\fR or \fBDELETE\fR command. The value is sent as an interger.
.IP /command...
The API command.
The command uses slash (\fB/\fR) to seperate the verbs in the same way that the CloudFlare v4 API documentation does.
Any verb starting with colon (\fB:\fR) is either converted to zone_id, user_id, organtization_id, or otherwise.

.SH RESULTS
The output is either JSON or YAML formatted.

.SH EXAMPLES
.B cli4 /zones
List infomation for all zones.

.B cli4 /zones/:example.com
List specific zone info.

.B cli4 /zones/:example.com/settings
List settings for a specific zone.

.B cli4 --delete purge_everything=true /zones/:example.com/purge_cache
Purge cache for a specific zone.

.B cli4 --delete files='[http://example.com/css/styles.css]' /zones/:example.com/purge_cache
Purge cache for a specific zone.

.B cli4 --delete files='[http://example.com/css/styles.css,http://example.com/js/script.js] /zones/:example.com/purge_cache
Purge cache for a specific zone.

.B cli4 --delete tags='[tag1,tag2,tag3]' /zones/:example.com/purge_cache
Purge cache for a specific zone.

.B cli4 /zones/:example.com/available_plans
List available plans for a zone.

.B cli4 --patch status=active /zones/:example.com/dnssec
Make DNSSEC active for specfic zone.

.B cli4 /zones/:example.com/dnssec
List DNSSEC infomation and status for a specific zone.

.SH SEE ALSO
The CloudFlare API can be found https://api.cloudflare.com/. Each API call is provided via a similarly named function within the CloudFlare class.


CHANGELOG.md
# Change Log

 - 2016-10-30 16:18:23 -0700 [2008a53](https://github.com/cloudflare/python-cloudflare/commit/2008a534891191ca37154f78aed1fd7de18aa5e6) flushed out tabs, converted print to print() - all python3 stuff
 - 2016-10-29 12:50:28 -0700 [e179521](https://github.com/cloudflare/python-cloudflare/commit/e1795216aea9935b19ab2340cd63f68a8ed4065e) Added more examples of using raw mode and updated README
 - 2016-10-27 09:15:22 -0700 [8f7b510](https://github.com/cloudflare/python-cloudflare/commit/8f7b51064c54884c095f598f9f93107a24cf2b7f) CHANGELOG.md creation added to Makefile - and pushed to github
 - 2016-10-27 09:09:47 -0700 [7711452](https://github.com/cloudflare/python-cloudflare/commit/7711452b94aa1cc556fef7e7430bcb005f7ca761) CHANGELOG.md creation added to Makefile
 - 2016-10-25 05:14:20 -0700 [206a699](https://github.com/cloudflare/python-cloudflare/commit/206a6992005fdc891e60bdf298e330d6de701810) working CTM release
 - 2016-10-25 05:10:25 -0700 [f107f40](https://github.com/cloudflare/python-cloudflare/commit/f107f40ad4388ad6f71c2e171d3b1a2eb9ba878a) support for CTM errors, cleaner exceptions for invalid methods, clearer cert exception message
 - 2016-10-25 05:06:08 -0700 [e0336e1](https://github.com/cloudflare/python-cloudflare/commit/e0336e1172029e0f25ffc5924edf8fbe86bb324f) typo
 - 2016-10-24 06:52:00 -0700 [1555a85](https://github.com/cloudflare/python-cloudflare/commit/1555a854655a3f4ca68dbcc7b4b46c72cb7451e9) typo
 - 2016-10-22 11:03:26 -0700 [b5919ad](https://github.com/cloudflare/python-cloudflare/commit/b5919ad53dcd9a992879004eb82758bee1bd3134) typo
 - 2016-10-22 10:58:46 -0700 [b0e6bb8](https://github.com/cloudflare/python-cloudflare/commit/b0e6bb8fdc0b7af5af726d2227acbdeaec3d7d41) add CTM (Cloudflare Traffic Manager) API calls
 - 2016-10-21 17:29:47 -0700 [406fa56](https://github.com/cloudflare/python-cloudflare/commit/406fa56b8666ddee1082f1bac0b7ade527c39037) create zone and populate example needed FQDN for CNAME dns record
 - 2016-10-21 14:08:43 -0700 [bc99cb5](https://github.com/cloudflare/python-cloudflare/commit/bc99cb513913ccc421543111d4631e1fa96fcebe) needed for package reasons
 - 2016-10-21 14:06:03 -0700 [460d77b](https://github.com/cloudflare/python-cloudflare/commit/460d77bad7b255718e3986090b75458c6eef49a0) 1.3.0 - python3 finally works via pip install
 - 2016-10-21 12:54:22 -0700 [110870b](https://github.com/cloudflare/python-cloudflare/commit/110870b4524cad34f7ddd780d9532f5b3e09d374) 1.2.6 - python3 does not need the import statement
 - 2016-10-21 12:49:37 -0700 [968b93c](https://github.com/cloudflare/python-cloudflare/commit/968b93c301b51c6d44a1976d06164d518af49e11) 1.2.5 - error_chain can now be read when an error happens
 - 2016-10-21 12:48:11 -0700 [3650b3e](https://github.com/cloudflare/python-cloudflare/commit/3650b3e841acb2e38b55722eb76e703a789a8d22) error_chain can now be read when an error happens
 - 2016-10-20 18:12:24 -0700 [5f78fcd](https://github.com/cloudflare/python-cloudflare/commit/5f78fcd803c4e314de41816caba9144e1f0c8d81) added examples and man page to distributed files
 - 2016-10-19 14:29:27 -0700 [c6c3175](https://github.com/cloudflare/python-cloudflare/commit/c6c3175543cfb11fa49a481565f7d8bfe3a22c65) fixed CloudFlare.exceptions in raise/except
 - 2016-10-18 18:16:34 -0700 [5c59702](https://github.com/cloudflare/python-cloudflare/commit/5c5970256744cc711ce408e5e8d7b12cd48a1ee3) added cli4 --dump documentation
 - 2016-10-18 17:42:50 -0700 [f89c64b](https://github.com/cloudflare/python-cloudflare/commit/f89c64b485e98366a0a942300cd26035f5931892) added more api calls - available_rate_plans  websockets ssl/analyze ssl/verification
 - 2016-10-18 17:38:28 -0700 [6655fee](https://github.com/cloudflare/python-cloudflare/commit/6655fee7ecd5dadfaf6abe4831ff39d5d72a253e) walk the tree correctly for --dump option
 - 2016-10-18 14:38:42 -0700 [3e177cf](https://github.com/cloudflare/python-cloudflare/commit/3e177cf0a4d29741822284980e46d93a143ff3ff) typo
 - 2016-10-17 17:19:26 -0700 [88c4250](https://github.com/cloudflare/python-cloudflare/commit/88c4250c40bd0ae43a6dcb36fe635b03080a0e31) typo
 - 2016-10-17 17:16:49 -0700 [d9f5930](https://github.com/cloudflare/python-cloudflare/commit/d9f593080a94178e9e703df1b78ab25af2e61c4d) added -d/--dump command to display list of API calls
 - 2016-10-17 17:16:15 -0700 [22b85b1](https://github.com/cloudflare/python-cloudflare/commit/22b85b1ba22e686792b03e053caeaaa8258d7008) made add internal call start with _ and added a few missing API calls
 - 2016-10-17 11:18:42 -0700 [a11c251](https://github.com/cloudflare/python-cloudflare/commit/a11c251c49253910cceac50286b4482934fc1807) first public raw mode release
 - 2016-10-17 11:16:45 -0700 [2dfe728](https://github.com/cloudflare/python-cloudflare/commit/2dfe728f42947a6ed0f126329bcb52461d35c54d) typo - theres just "result" and "result_info" returned in raw mode
 - 2016-10-17 11:07:17 -0700 [797b90f](https://github.com/cloudflare/python-cloudflare/commit/797b90f11889bca2be70d923ef31e139b0dded91) Added --raw mode examples
 - 2016-10-17 10:38:39 -0700 [c1ead41](https://github.com/cloudflare/python-cloudflare/commit/c1ead419110b5e50bda69061ca2b8635eae085eb) Added more cleanup code and sdist/bdist options - not that bdist is useful
 - 2016-10-16 19:05:36 -0700 [06e5fcc](https://github.com/cloudflare/python-cloudflare/commit/06e5fccff0ed2071b851f66dab10ed2e6822554e) with raw code added we bump the version number
 - 2016-10-16 19:04:59 -0700 [052e1c0](https://github.com/cloudflare/python-cloudflare/commit/052e1c046bd36fe57dffaaefab7a430a207f094f) example of how to page thru data with raw option
 - 2016-10-16 18:13:28 -0700 [3ae4626](https://github.com/cloudflare/python-cloudflare/commit/3ae4626c5703470fd50b00066af669f777d40646) added raw flag to class so that paging values can be returned
 - 2016-10-16 18:11:59 -0700 [00ef9f8](https://github.com/cloudflare/python-cloudflare/commit/00ef9f8396c067d30c861b68336686d5680887a1) typo

cli4/cli4.py
#!/usr/bin/env python
"""Cloudflare API via command line"""

import os
import sys
import re
import getopt
import json
try:
    import yaml
except ImportError:
    yaml = None

sys.path.insert(0, os.path.abspath('..'))
import CloudFlare
import CloudFlare.exceptions

def convert_zones_to_identifier(cf, zone_name):
    """zone names to numbers"""
    params = {'name':zone_name, 'per_page':1}
    try:
        zones = cf.zones.get(params=params)
    except CloudFlare.exceptions.CloudFlareAPIError as e:
        exit('cli4: %s - %d %s' % (zone_name, e, e))
    except Exception as e:
        exit('cli4: %s - %s' % (zone_name, e))

    for zone in zones:
        if zone_name == zone['name']:
            return zone['id']

    exit('cli4: %s - zone not found' % (zone_name))

def convert_dns_record_to_identifier(cf, zone_id, dns_name):
    """dns record names to numbers"""
    # this can return an array of results as there can be more than one DNS entry for a name.
    params = {'name':dns_name}
    try:
        dns_records = cf.zones.dns_records.get(zone_id, params=params)
    except CloudFlare.exceptions.CloudFlareAPIError as e:
        exit('cli4: %s - %d %s' % (dns_name, e, e))
    except Exception as e:
        exit('cli4: %s - %s' % (dns_name, e))

    r = []
    for dns_record in dns_records:
        if dns_name == dns_record['name']:
            r.append(dns_record['id'])
    if len(r) > 0:
        return r

    exit('cli4: %s - dns name not found' % (dns_name))

def convert_certificates_to_identifier(cf, certificate_name):
    """certificate names to numbers"""
    try:
        certificates = cf.certificates.get()
    except CloudFlare.exceptions.CloudFlareAPIError as e:
        exit('cli4: %s - %d %s' % (certificate_name, e, e))
    except Exception as e:
        exit('cli4: %s - %s' % (certificate_name, e))

    for certificate in certificates:
        if certificate_name in certificate['hostnames']:
            return certificate['id']

    exit('cli4: %s - no zone certificates found' % (certificate_name))

def convert_organizations_to_identifier(cf, organization_name):
    """organizations names to numbers"""
    try:
        organizations = cf.user.organizations.get()
    except CloudFlare.exceptions.CloudFlareAPIError as e:
        exit('cli4: %s - %d %s' % (organization_name, e, e))
    except Exception as e:
        exit('cli4: %s - %s' % (organization_name, e))

    for organization in organizations:
        if organization_name == organization['name']:
            return organization['id']

    exit('cli4: %s - no organizations found' % (organization_name))

def convert_invites_to_identifier(cf, invite_name):
    """invite names to numbers"""
    try:
        invites = cf.user.invites.get()
    except CloudFlare.exceptions.CloudFlareAPIError as e:
        exit('cli4: %s - %d %s' % (invite_name, e, e))
    except Exception as e:
        exit('cli4: %s - %s' % (invite_name, e))

    for invite in invites:
        if invite_name == invite['organization_name']:
            return invite['id']

    exit('cli4: %s - no invites found' % (invite_name))

def convert_virtual_dns_to_identifier(cf, virtual_dns_name):
    """virtual dns names to numbers"""
    try:
        virtual_dnss = cf.user.virtual_dns.get()
    except CloudFlare.exceptions.CloudFlareAPIError as e:
        exit('cli4: %s - %d %s\n' % (virtual_dns_name, e, e))
    except Exception as e:
        exit('cli4: %s - %s\n' % (virtual_dns_name, e))

    for virtual_dns in virtual_dnss:
        if virtual_dns_name == virtual_dns['name']:
            return virtual_dns['id']

    exit('cli4: %s - no virtual_dns found' % (virtual_dns_name))

def walk(m, s):
    """recursive walk of the tree"""
    for n in sorted(dir(m)):
        if n[0] == '_':
            # internal
            continue
        if n in ['delete', 'get', 'patch', 'post', 'put']:
            # gone too far
            continue
        a = getattr(m, n)
        d = dir(a)
        if '_base' in d:
            # it's a known api call - lets show the result and continue down the tree
            if 'delete' in d or 'get' in d or 'patch' in d or 'post' in d or 'put' in d:
                # only show the result if a call exists for this part
                print(s + '/' + n)
            walk(a, s + '/' + n)

def dump_commands(cf):
    """dump a tree of all the known API commands"""
    walk(cf, '')

def cli4(args):
    """Cloudflare API via command line"""

    verbose = False
    output = 'json'
    raw = False
    dump = False
    method = 'GET'

    usage = ('usage: cli4 '
             + '[-V|--version] [-h|--help] [-v|--verbose] [-q|--quiet] [-j|--json] [-y|--yaml] '
             + '[-r|--raw] '
             + '[-d|--dump] '
             + '[--get|--patch|--post|-put|--delete] '
             + '[item=value ...] '
             + '/command...')

    try:
        opts, args = getopt.getopt(args,
                                   'VhvqjyrdGPOUD',
                                   [
                                       'version',
                                       'help', 'verbose', 'quiet', 'json', 'yaml',
                                       'raw',
                                       'dump',
                                       'get', 'patch', 'post', 'put', 'delete'
                                   ])
    except getopt.GetoptError:
        exit(usage)
    for opt, arg in opts:
        if opt in ('-V', '--version'):
            exit('Cloudflare library version: %s' % (CloudFlare.__version__))
        if opt in ('-h', '--help'):
            exit(usage)
        elif opt in ('-v', '--verbose'):
            verbose = True
        elif opt in ('-q', '--quiet'):
            output = None
        elif opt in ('-j', '--json'):
            output = 'json'
        elif opt in ('-y', '--yaml'):
            output = 'yaml'
        elif opt in ('-r', '--raw'):
            raw = True
        elif opt in ('-d', '--dump'):
            dump = True
        elif opt in ('-G', '--get'):
            method = 'GET'
        elif opt in ('-P', '--patch'):
            method = 'PATCH'
        elif opt in ('-O', '--post'):
            method = 'POST'
        elif opt in ('-U', '--put'):
            method = 'PUT'
        elif opt in ('-D', '--delete'):
            method = 'DELETE'

    digits_only = re.compile('^[0-9]+$')

    # next grab the params. These are in the form of tag=value
    params = {}
    while len(args) > 0 and '=' in args[0]:
        tag_string, value_string = args.pop(0).split('=', 1)
        if value_string == 'true':
            value = True
        elif value_string == 'false':
            value = False
        elif value_string[0] is '=' and digits_only.match(value_string[1:]):
            value = int(value_string[1:])
        elif value_string[0] in '[{' and value_string[-1] in '}]':
            # a json structure - used in pagerules
            try:
                #value = json.loads(value) - changed to yaml code to remove unicode string issues
                if yaml is None:
                    exit('cli4: install yaml support')
                value = yaml.safe_load(value_string)
            except ValueError:
                exit('cli4: %s="%s" - can\'t parse json value' % (tag_string, value_string))
        else:
            value = value_string
        tag = tag_string
        params[tag] = value

    if dump:
        cf = CloudFlare.CloudFlare()
        dump_commands(cf)
        exit(0)

    # what's left is the command itself
    if len(args) != 1:
        exit(usage)

    command = args[0]
    # remove leading and trailing /'s
    if command[0] == '/':
        command = command[1:]
    if command[-1] == '/':
        command = command[:-1]

    # break down command into it's seperate pieces
    # these are then checked against the Cloudflare class
    # to confirm there is a method that matches
    parts = command.split('/')

    cmd = []
    identifier1 = None
    identifier2 = None

    hex_only = re.compile('^[0-9a-fA-F]+$')

    cf = CloudFlare.CloudFlare(debug=verbose, raw=raw)

    m = cf
    for element in parts:
        if element[0] == ':':
            element = element[1:]
            if identifier1 is None:
                if len(element) in [32, 40, 48] and hex_only.match(element):
                    # raw identifier - lets just use it as-is
                    identifier1 = element
                elif cmd[0] == 'certificates':
                    # identifier1 = convert_certificates_to_identifier(cf, element)
                    identifier1 = convert_zones_to_identifier(cf, element)
                elif cmd[0] == 'zones':
                    identifier1 = convert_zones_to_identifier(cf, element)
                elif cmd[0] == 'organizations':
                    identifier1 = convert_organizations_to_identifier(cf, element)
                elif (cmd[0] == 'user') and (cmd[1] == 'organizations'):
                    identifier1 = convert_organizations_to_identifier(cf, element)
                elif (cmd[0] == 'user') and (cmd[1] == 'invites'):
                    identifier1 = convert_invites_to_identifier(cf, element)
                elif (cmd[0] == 'user') and (cmd[1] == 'virtual_dns'):
                    identifier1 = convert_virtual_dns_to_identifier(cf, element)
                else:
                    exit("/%s/%s :NOT CODED YET 1" % ('/'.join(cmd), element))
                cmd.append(':' + identifier1)
            else:
                if len(element) in [32, 40, 48] and hex_only.match(element):
                    # raw identifier - lets just use it as-is
                    identifier2 = element
                elif (cmd[0] and cmd[0] == 'zones') and (cmd[2] and cmd[2] == 'dns_records'):
                    identifier2 = convert_dns_record_to_identifier(cf, identifier1, element)
                else:
                    exit("/%s/%s :NOT CODED YET 2" % ('/'.join(cmd), element))
                # identifier2 may be an array - this needs to be dealt with later
                if isinstance(identifier2, list):
                    cmd.append(':' + '[' + ','.join(identifier2) + ']')
                else:
                    cmd.append(':' + identifier2)
                    identifier2 = [identifier2]
        else:
            try:
                m = getattr(m, element)
                cmd.append(element)
            except AttributeError:
                # the verb/element was not found
                if len(cmd) == 0:
                    exit('cli4: /%s - not found' % (element))
                else:
                    exit('cli4: /%s/%s - not found' % ('/'.join(cmd), element))

    results = []
    if identifier2 is None:
        identifier2 = [None]
    for i2 in identifier2:
        try:
            if method is 'GET':
                r = m.get(identifier1=identifier1, identifier2=i2, params=params)
            elif method is 'PATCH':
                r = m.patch(identifier1=identifier1, identifier2=i2, data=params)
            elif method is 'POST':
                r = m.post(identifier1=identifier1, identifier2=i2, data=params)
            elif method is 'PUT':
                r = m.put(identifier1=identifier1, identifier2=i2, data=params)
            elif method is 'DELETE':
                r = m.delete(identifier1=identifier1, identifier2=i2, data=params)
            else:
                pass
        except CloudFlare.exceptions.CloudFlareAPIError as e:
            if len(e) > 0:
                # more than one error returned by the API
                for x in e:
                    sys.stderr.write('cli4: /%s - %d %s\n' % (command, x, x))
            exit('cli4: /%s - %d %s' % (command, e, e))
        except Exception as e:
            exit('cli4: /%s - %s - api error' % (command, e))

        results.append(r)

    if len(results) == 1:
        results = results[0]

    if output == 'json':
        print(json.dumps(results, indent=4, sort_keys=True))
    if output == 'yaml' and yaml is not None:
        print(yaml.safe_dump(results))

